pkg/operator/cockroachdb/controller.go

```diff
@@ -0,0 +1,573 @@
+				"prometheus.io/path":                                     "_status/vars",
+				"prometheus.io/port":                                     strconv.Itoa(int(httpPort)),
+			},
+		},
+		Spec: v1.ServiceSpec{
+			Selector: createAppLabels(),
+			// We want all pods in the StatefulSet to have their addresses published for
+			// the sake of the other CockroachDB pods even before they're ready, since they
+			// have to be able to talk to each other in order to become ready.
+			PublishNotReadyAddresses: true,
+			ClusterIP:                "None",
+			Ports:                    createServicePorts(httpPort, grpcPort),
+		},
+	}
+
+	if _, err := c.context.Clientset.CoreV1().Services(cluster.namespace).Create(replicaService); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return err
+		}
+		logger.Infof("replica service %s already exists in namespace %s", replicaService.Name, replicaService.Namespace)
+	} else {
+		logger.Infof("replica service %s started in namespace %s", replicaService.Name, replicaService.Namespace)
+	}
+
+	return nil
+}
+
+func (c *ClusterController) createPodDisruptionBudget(cluster *cluster) error {
+	maxUnavailable := intstr.FromInt(int(1))
+
+	pdb := &policyv1beta1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "cockroachdb-budget",
+			Namespace:       cluster.namespace,
+			OwnerReferences: []metav1.OwnerReference{cluster.ownerRef},
+			Labels:          createAppLabels(),
+		},
+		Spec: policyv1beta1.PodDisruptionBudgetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: createAppLabels(),
+			},
+			MaxUnavailable: &maxUnavailable,
+		},
+	}
+
+	if _, err := c.context.Clientset.PolicyV1beta1().PodDisruptionBudgets(cluster.namespace).Create(pdb); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return err
+		}
+		logger.Infof("pod disruption budget %s already exists in namespace %s", pdb.Name, pdb.Namespace)
+	} else {
+		logger.Infof("pod disruption budget %s created in namespace %s", pdb.Name, pdb.Namespace)
+	}
+
+	return nil
+}
+
+func (c *ClusterController) createStatefulSet(cluster *cluster) error {
+	replicas := int32(cluster.spec.Storage.NodeCount)
+
+	httpPort, grpcPort, err := getPortsFromSpec(cluster.spec.Network)
+	if err != nil {
+		return err
+	}
+
+	var envVarChannelVal string
+	if cluster.spec.Secure {
+		envVarChannelVal = envVarValChannelSecure
+	} else {
+		envVarChannelVal = envVarValChannelInsecure
+	}
+
+	terminationGracePeriodSeconds := int64(60)
+
+	statefulSet := &appsv1beta1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            appName,
+			Namespace:       cluster.namespace,
+			OwnerReferences: []metav1.OwnerReference{cluster.ownerRef},
+		},
+		Spec: appsv1beta1.StatefulSetSpec{
+			ServiceName: appName,
+			Replicas:    &replicas,
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: cluster.namespace,
+					Labels:    createAppLabels(),
+				},
+				Spec: v1.PodSpec{
```

Comment:
This is getting nested to the point where it's hard to read. Can you break this out into its own function to alleviate some of that?

Reply:
Sure i can break up the collection of k8s resources here so the nesting isn't so deep.